stress/src/async_throughput.rs

```diff
@@ -0,0 +1,257 @@
+        {
+            let stop: Arc<AtomicBool> = self.stop.clone();
+            tokio::spawn(async move {
+                tokio::signal::ctrl_c().await.ok();
+                println!("\n\nGraceful shutdown initiated...");
+                stop.store(true, Ordering::SeqCst);
+            });
+        }
+
+        // Stats reporter
+        let reporter = self.spawn_reporter();
+
+        // Worker tasks
+        let mut worker_handles = Vec::new();
+        for worker_id in 0..workers {
+            let handle = self.spawn_worker(worker_id, concurrency_per_worker, test_fn.clone());
+            worker_handles.push(handle);
+        }
+
+        // Wait for all workers
+        let worker_results = futures::future::join_all(worker_handles).await;
+
+        // Stop reporter
+        reporter.abort();
+
+        // Print final stats
+        self.print_final_stats(test_start.elapsed(), worker_results)
+            .await;
+
+        Ok(())
+    }
+
+    fn spawn_reporter(&self) -> tokio::task::JoinHandle<()> {
+        let stats = self.stats.clone();
+        let stop = self.stop.clone();
+
+        tokio::spawn(async move {
+            let mut interval = tokio::time::interval(Duration::from_secs(1));
+            let mut last_success = 0u64;
+
+            loop {
+                interval.tick().await;
+
+                if stop.load(Ordering::SeqCst) {
+                    break;
+                }
+
+                let success = stats.success_count.load(Ordering::Relaxed);
+                let errors = stats.error_count.load(Ordering::Relaxed);
+                let success_delta = success - last_success;
+                if success_delta > 0 || errors > 0 {
+                    let throughput = success_delta;
+
+                    println!(
+                        "Throughput: {:>7}/s | Total: {:>8} | Errors: {}",
+                        throughput.to_formatted_string(&Locale::en),
+                        success.to_formatted_string(&Locale::en),
+                        errors
+                    );
+                }
+
+                last_success = success;
+            }
+        })
+    }
+
+    fn spawn_worker<F, Fut, E>(
+        &self,
+        worker_id: usize,
+        concurrency: usize,
+        test_fn: Arc<F>,
+    ) -> tokio::task::JoinHandle<(usize, u64, u64)>
+    where
+        F: Fn() -> Fut + Send + Sync + 'static,
+        Fut: std::future::Future<Output = Result<(), E>> + Send + 'static,
+        E: std::error::Error + Send + Sync + 'static,
+    {
+        let stats = self.stats.clone();
+        let stop = self.stop.clone();
+
+        tokio::spawn(async move {
+            let mut futures = FuturesUnordered::new();
+            let mut local_success = 0u64;
+            let mut local_errors = 0u64;
+
+            // Main loop
+            while !stop.load(Ordering::SeqCst) {
+                // Keep pipeline full
+                while futures.len() < concurrency && !stop.load(Ordering::SeqCst) {
```

Comment:
I think using [`buffer_unordered`](https://docs.rs/futures/latest/futures/stream/trait.StreamExt.html#method.buffer_unordered) with the required concurrency is much better than manually maintaining concurrency.

Reply:
Good point. done.